### DISCUSSION TITLE
GSoC 2022: Storage API

### DISCUSSION THREAD

**Discussion by MariamFahmy98** [pgmoneta/pgmoneta#54] [Ideas]
Hi,

I am Mariam Fahmy, I have been selected for [```GSoC 2022```](https://summerofcode.withgoogle.com/programs/2022/projects/EpwXhnaa) as a contributor to the pgmoneta project.

I will be working on implementing various storage engines, I would like to complete the following work:

1. Implementing a local storage engine:
    It will have a minimal implementation since **pg_basebackup** already creates a backup for the server locally. 

2. Implementing a SSH storage engine:
    **libssh** will be used to implement the SSH storage engine, specifically the **SFTP api** to copy files and directories to the remote system, in addition to deleting backup remote directories.

3. Implementing Amazon S3 storage engine:
    S3 API of Ceph.io will be used to implement the S3 storage engine, It is compatible with the basic data access model of the Amazon S3 API. 
In case the basic operations of ceph.io don’t cover the required use cases, HTTP and XML functionality will be added to the project such that Amazon S3 storage engine mainly will use:
    - HTTP (http.c & http.h): Its main functionality is to create a HTTP header by adding new header attributes, and sending a HTTP request to Amazon S3 REST api using libcurl.
    - XML (xml.c & xml.h): Its main functionality is to create a HTTP body by adding a XML element with the required data to make a HTTP request.

For more details, you can check out the main proposal:
https://docs.google.com/document/d/1EbRzgfZCDWG6LCD0puil8bUt10aehT4skfZ1YFlbNKk/edit?usp=sharing

I am open to new suggestions and ideas, feel free to either discuss it with me here or through email. you can contact me at ```mariamfahmy66@gmail.com```.

Looking forward to an amazing learning experience.

**Here is a list of the work done:**

1. Local Storage Engine:

- Related Commit: [Adding a local storage workflow](https://github.com/pgmoneta/pgmoneta/commit/55b1e24f73ca0c4f191927df546f43b37c41f06d)

2. Splitting commands into workflows:

-    Related issue: [Splitting commands into workflows](https://github.com/pgmoneta/pgmoneta/issues/72)
-    Related Commits:
    [Adding restore workflow](https://github.com/pgmoneta/pgmoneta/commit/3ef1784d7391c26e6426d8c52ddaa41e8d0ad034)
    [Adding archive workflow](https://github.com/pgmoneta/pgmoneta/commit/ad488d03eab88e26ebb2cb1189c65ea0688205cd)
    [Creating a delete backup workflow](https://github.com/pgmoneta/pgmoneta/commit/59634713eee93e286b6ac4db36324491acb92d60)
    [Adding a retention workflow](https://github.com/pgmoneta/pgmoneta/commit/ebf9e8075223f568991d774c56c38b98b704c9a7)

3. SSH Storage Engine:

- Related issue: [Implementing SSH Storage Engine](https://github.com/pgmoneta/pgmoneta/issues/77)
- Related Commits:
  [Adding a remote backup](https://github.com/pgmoneta/pgmoneta/commit/b299929e0fb8f85d66159ec91e0bd1d9bcc08452)
  [Adding SSH link functionality](https://github.com/pgmoneta/pgmoneta/commit/c9aebe875851a0cd679d41c22f2ec5d9d54d98f1)

4. Amazon S3 Storage Engine:

- Related issue: [Implementing AWS S3 storage engine](https://github.com/pgmoneta/pgmoneta/issues/97)
- Related Commit: [Supporting backup in AWS S3 engine](https://github.com/pgmoneta/pgmoneta/commit/a5330ee2ea9c30afa8742141e02ceafd71bb86d2)

5. Azure Storage Engine:

- Related issue: [Implementing Azure Storage Engine](https://github.com/pgmoneta/pgmoneta/issues/100)
- Related Commit: [Supporting backup in Azure storage account](https://github.com/pgmoneta/pgmoneta/commit/a4d57866879ca47468cf379ac7bd8364163bb4fd)

Future Work:
Supporting other pgmoneta-cli commands as restore, archive, delete and retain in:
- SSH Storage Engine
- AWS S3 Storage Engine
- Azure Storage Engine

**Comment by jesperpedersen** [comment]
Welcome !

I think this is a solid proposal for the work ahead, and starting out with the local engine should give us an idea of the interface needed - even though the simplest of them all.

For 3) -- ideally we don't want any S3 specific code in our repository but reuse the Ceph API for this. So, if it is necessary then we need to create feedback to Ceph upstream with our missing requirements such that we can remove our code at a later time.

I think the baseline requirements for the dependencies are defined by RHEL8 - even though RHEL9 is out...

Feel free to link any pull requests - in any form - in this thread such that we can get a better idea of how to proceed. The https://github.com/pgmoneta/pgmoneta/blob/main/CONTRIBUTING.md guide should provide a starting point.

Thanks !

**Comment by MariamFahmy98** [comment]
**Here is a list of the work done:**

1. Local Storage Engine:

- Related Commit: [Adding a local storage workflow](https://github.com/pgmoneta/pgmoneta/commit/55b1e24f73ca0c4f191927df546f43b37c41f06d)

2. Splitting commands into workflows:

-    Related issue: [Splitting commands into workflows](https://github.com/pgmoneta/pgmoneta/issues/72)
-    Related Commits:
    [Adding restore workflow](https://github.com/pgmoneta/pgmoneta/commit/3ef1784d7391c26e6426d8c52ddaa41e8d0ad034)
    [Adding archive workflow](https://github.com/pgmoneta/pgmoneta/commit/ad488d03eab88e26ebb2cb1189c65ea0688205cd)
    [Creating a delete backup workflow](https://github.com/pgmoneta/pgmoneta/commit/59634713eee93e286b6ac4db36324491acb92d60)
    [Adding a retention workflow](https://github.com/pgmoneta/pgmoneta/commit/ebf9e8075223f568991d774c56c38b98b704c9a7)

3. SSH Storage Engine:

- Related issue: [Implementing SSH Storage Engine](https://github.com/pgmoneta/pgmoneta/issues/77)
- Related Commits:
  [Adding a remote backup](https://github.com/pgmoneta/pgmoneta/commit/b299929e0fb8f85d66159ec91e0bd1d9bcc08452)
  [Adding SSH link functionality](https://github.com/pgmoneta/pgmoneta/commit/c9aebe875851a0cd679d41c22f2ec5d9d54d98f1)

4. Amazon S3 Storage Engine:

- Related issue: [Implementing AWS S3 storage engine](https://github.com/pgmoneta/pgmoneta/issues/97)
- Related Commit: [Supporting backup in AWS S3 engine](https://github.com/pgmoneta/pgmoneta/commit/a5330ee2ea9c30afa8742141e02ceafd71bb86d2)

5. Azure Storage Engine:

- Related issue: [Implementing Azure Storage Engine](https://github.com/pgmoneta/pgmoneta/issues/100)
- Related Commit: [Supporting backup in Azure storage account](https://github.com/pgmoneta/pgmoneta/commit/a4d57866879ca47468cf379ac7bd8364163bb4fd)

Future Work:
Supporting other pgmoneta-cli commands as restore, archive, delete and retain in:
- SSH Storage Engine
- AWS S3 Storage Engine
- Azure Storage Engine

**Comment by HazemRawi** [comment]
Hello,

Currently WAL shipped to local directories only ,I want to support shipping WAL segments to a remote storage engine (SSH,S3 and Azure)
As

`      ev_periodic_init (&wal_streaming, wal_streaming_cb, 0., 60, 0);`

makes wal streaming starts automatically every 60 second 

Should I change the shipping directory in `pgmoneta_wal()` to choose between different storage engines to make shipping WAL segment to remote servers possible?

**Reply by HazemRawi** [reply]
This will also be managed in `se_ssh` , `se_s3` and `se_azure`

**Reply by jesperpedersen** [reply]
`storage_engine` should be a comma separated list, so if `local, ssh` then both "accounts" should be used.

Everything should be transparent to the user if the configuration is correct

**Reply by jesperpedersen** [reply]
WAL streaming is in `wal.c` so you need to integrate into that file

**Reply by HazemRawi** [reply]
> `storage_engine` should be a comma separated list, so if `local, ssh` then both "accounts" should be used.

`struct configuration` now support only one storage engine using 

`   int storage_engine;  /**< The storage engine */`

So I think it is better to make it array of integers to store different types of storage engines inside it,
considering that this approach will require some modifications [here](https://github.com/pgmoneta/pgmoneta/blob/4b816e1f77fd22cf0c12d19e0ad6d977ebe1226a/src/libpgmoneta/workflow.c#L149-L168).

**Reply by HazemRawi** [reply]
Another approach is to keep `storage_engine` and assign each storage engine type (local, SSH, S3, and Azure) a flag bit.

    Local: 2^0
    SSH: 2^1
    S3: 2^2
    Azure: 2^3

This way, when we want to check for existence for a storage engine type we can:

`   if (config->storage_engine & STORAGE_ENGINE_SSH)`

I prefer this approach over using an array of integers. Nevertheless, I am waiting for your feedback.

**Reply by jesperpedersen** [reply]
Making flags is a good solution

**Reply by HazemRawi** [reply]
I've to create `wal_open()` function for ssh storage engine
so where should I place it?
in `storage.h` and give it a name `pgmoneta_stfp_wal_open()` -- global function
or in `wal.c` and name it `sftp_wal_open()` -- static function

**Reply by HazemRawi** [reply]
Also, I will handle the whole shipping process using `sftp_write()`
so WAL shipping to ssh se will done simultaneously with shipping to local directory.

**Reply by jesperpedersen** [reply]
As much as possible should be in `se_ssh.c`, so you need to check the flag in `wal.c` and then call functions in `se_ssh.c`.

I would start a work branch, and then update this thread once you have done some functionality that head in this direction. That way it is easier to figure out the best design. I think

```
 if (config->storage_engine & STORAGE_ENGINE_SSH)
```

is a good idea, and we have to see where the "pieces" fit - not all can be in `workflow.c`

**Comment by HazemRawi** [comment]
Hello,
I am currently working on the restore command to support restoring backups from an SSH server. As part of this feature, should I make it optionally look for a backup on a specific storage engine, or should it search through all storage engines to find the desired backup?
it will be something like 
`    pgmoneta-cli -c pgmoneta.conf restore primary newest SSH current /tmp/`

**Reply by jesperpedersen** [reply]
Right now, the storage engine is defined at configuration level - we can make that server level as well later.

 So, you know where the backups are located - so no need for the `SSH` option

**Comment by HazemRawi** [comment]
Hello,

I’m currently working on support functions for the SSH storage engine, and I noticed that when Mariam implemented `sftp_copy_directory` function, she used standard `opendir`, `readdir`, and `DIR` instead of `sftp_opendir` , `sftp_readdir` and `sftp_dir`.

I wanted to understand if there are any advantages to using the `sftp_` prefixed functions in this context. Personally, I would prefer to use the sftp_ functions to maintain consistency with the SSH storage engine.

**Reply by jesperpedersen** [reply]
If you mean https://github.com/pgmoneta/pgmoneta/blob/main/src/libpgmoneta/se_ssh.c#L545 then it copies from a local directory to the remote directory - so `opendir()` is ok here.

A backup is done locally, and then transferred
